src/app.rs

```diff
@@ -12,78 +12,133 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::identity::SecretManager;
-use crate::metrics::Metrics;
-use crate::state::ProxyStateManager;
-use crate::{admin, config, identity, proxy, readiness, signal, stats};
-use anyhow::Context;
-use prometheus_client::registry::Registry;
 use std::net::SocketAddr;
 use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
 use std::thread;
 use std::time::Duration;
+
+use anyhow::Context;
+use prometheus_client::registry::Registry;
 use tracing::Instrument;
 
+use crate::dns;
+use crate::identity::SecretManager;
+use crate::metrics::Metrics;
+use crate::state::ProxyStateManager;
+use crate::{admin, config, identity, metrics, proxy, readiness, signal};
+
 pub async fn build_with_cert(
     config: config::Config,
     cert_manager: Arc<SecretManager>,
 ) -> anyhow::Result<Bound> {
-    let mut registry = Registry::default();
-    let metrics = Arc::new(Metrics::from(&mut registry));
-
     let shutdown = signal::Shutdown::new();
     // Setup a drain channel. drain_tx is used to trigger a drain, which will complete
     // once all drain_rx handlers are dropped.
-    // Any component which wants time to gracefully exit should take in a drain_rx clone, await drain_rx.signaled(), then cleanup.
+    // Any component which wants time to gracefully exit should take in a drain_rx clone,
+    // await drain_rx.signaled(), then cleanup.
     // Note: there is still a hard timeout if the draining takes too long
     let (drain_tx, drain_rx) = drain::channel();
 
+    // Register readiness tasks.
     let ready = readiness::Ready::new();
-    let proxy_task = ready.register_task("proxy listeners");
+    let state_mgr_task = ready.register_task("state manager");
+    let proxy_task = if config.proxy {
+        Some(ready.register_task("proxy"))
+    } else {
+        None
+    };
+    let dns_task = if config.dns_proxy {
+        Some(ready.register_task("dns proxy"))
+    } else {
+        None
+    };
+
+    // Create and start the readiness server.
+    let readiness_server = readiness::Server::new(config.clone(), drain_rx.clone(), ready.clone())
+        .await
+        .context("readiness server starts")?;
+    let readiness_address = readiness_server.address();
+    readiness_server.spawn();
```

Comment:
I think we need this in the data plane task pool, as discussed on slack

Reply:
My takeaway from our discussion is that this one was on the edge of control/data plane and didn't need to be in the data plane task pool?